.cargo/config.toml

```diff
@@ -0,0 +1,8 @@
+[target.'cfg(target_os = "none")']
```

Comment:
> This looks a bit similar to qboot-libkrunfw in principle, ...
Right. In principle, we try to create a small firmware for jump into a payload entry point - that is definitely our purpose.

> However, the code you presented seems quite a bit larger and more complex, ...
It is a bit larger and more complex, because we need below features (which is not available in qboot)
1) The guest entrypoint must be a guest reset vector at 0xFFFFFFF0 - That is required by TDX.
2) The reset vector must be 32bit protected mode, then switch to 64bit long mode with paging. - That is required by TDX.
3) TD-Shim need *accept* the memory assigned by the VMM. - That is required by TDX.
4) Any information from VMM (such as memory map) must be measured into a measurement register and provide an event log for attestation later. - That is required by TDX.
5) We need support multiple format of payload (it could be PE/COFF image, ELF image, or bzImage, vmlinux). The payload type is described at https://github.com/jyao1/td-shim/blob/init_version/doc/tdshim_spec.md#td-payload-info-guid-extension-hob. - That is a feature in Td-Shim.
6) We need enable best practice for defense in depth, such as data execution protection, stack guard, shadow stack. - That is a feature in Td-Shim.

> We don't really understand secure version number based attestation. 
That is OK. TDX based attestation is a big topic. I add in PPT, just to explain why we need secure boot. We have not enabled secure boot yet. I think we can discuss later. Maybe a dedicated session for TDX attestation.

> Is there a relationship with https://github.com/cloud-hypervisor/rust-hypervisor-firmware ?
I would say: We are inspired by this project - rust-hypervisor-firmware. It made us think that a small rust-based firmware is good choice for cloud hv and container use case.
In the beginning, we did leverage some code such as PE loader, or some initialization. But now, our major code is different due to special TDX requirement or new feature.

> There is a trend in firmware to split between low-level hardware bring up and firmware interfaces
Exactly. The concept is similar to coreboot. 
I have been working on firmware area since 2003. A clear R&R is our final goal. Even in EDKII project, we introduced  UefiPayloadPkg (https://github.com/tianocore/edk2/tree/master/UefiPayloadPkg).
In our context, payload is some td-shim can boot to or transfer to directly.
A payload image could be a linux kernel (in kernel direct boot use case).
A payload image could be a UEFI payload, which allows you run a UEFI environment, then boot UEFI OS.
A payload image may also be a bare metal environment running a single service (such as TLS session setup in TDX Migration TD use case)

> That split is less useful in virtualized systems, where the hardware bring up is simpler 
If it is simpler, I agree, you can bring up the system with Payload image directly.
I won't disagree if you want to do step 1~4 in your payload image, if you think it is simple.
I am a little confused now. On one hand, people think it is a bit large and complex. On the other hand, people think a virtualized system does not need split because it is simple.
So is this simple or complex? :-)

> This leads us to assume that the complexity in td-shim may come from other reasons, but the value added by this this shim (as documented in docs/design.md) is not super clear.
I describes the complexity due to special requirement introduced by TDX. I would say the value is
1) Current solution is EDKII based TDVF. But it is very big. Td-Shim provides a lightweight solution. (Just like rust-hypervisor-firmware/coreboot v.s. EDKII) - That is the main reason.
2) We use td-shim project to our service-td development. Having one single solution can help code maintenance. 
3) You can write your own small guest firmware to enable TDX. Not a problem. But TDX has some very special function requirement and security requirement (such as reset vector, mode switch, accept memory, measurement, event log, etc.). td-shim provide a reference.

> What are the requirements for the payload? 
The payload type is described at https://github.com/jyao1/td-shim/blob/init_version/doc/tdshim_spec.md#td-payload-info-guid-extension-hob.
td-shim just jumps to payload with defined interface. In payload, you can do whatever you want to do.
E.g. you can stay in the payload and provide service. Or you can load the next phase image, and jump to it.
That is out of td-shim's control. td-shim is just a shim-firmware to help you launch tdx environment.

> Does this support / require direct kernel boot?
It does support direct kernel boot. td-shim can boot a linux kernel.
Td-shim does not require direct kernel boot. You can provide you own payload image (e.g. an ELF or PE/COFF image).

Reply:
>However, the code you presented seems quite a bit larger and more complex (126 files, 8614 lines of Rust code, 1444 of assembly code). Some of this complexity may be related to TDX, but how much is unclear to the reviewers so far.

The "126 files" seems like a great feature to me:)
There are non-std-test framework, unit test cases, fuzz test, documentation, and great modularization. There are 8 .gitingore and 18 Cargo.toml:)

But great efforts may be needed due to:
1) asm code
2) code is still lack of comments, which measns we must be familiar with corresponding specifications to maintain the code.

Reply:
@jyao1 Thanks a lot for the detailed explanation.

I would suggest adding the "value" part to the documentation (where I put another comment). It's great as you explained it. I also think that your explanations about TDX requirements would enhance the documentation part that describes the TD interface.

Regarding our smiley-question "is it complex or simple", the point is that a lot of the complexity we see in regular firmware is not needed for virtualized hardware, so that if we see hardware-level complexity in something that is initially intended for virtual-only environments like TDX, and also is already split to some other payload, we wonder where the complexity comes from. I believe that you answered that by pointing out what is specifically required by TDX.

Thanks again.

I intend to continue my review today if I can find the time.

Reply:
@jiangliu Files are not a feature per se. 126 files is a lot for a minimalist firmware. It is expected for a full featured firmware. It would be quite low for an operating system. To be clear, when I reported that comment made by another engineer, it was simply in contrast to qboot. I believe that @jyao1 did answer my questions quite well, and that the answers should be recorded prominently in the documentation.

Reply:
@c3d This is a shim firmware for TDX, not a minimal (In the absolute sense), generic firmware. A full featured UEFI compliant firmware is a *lot* more than 126 files btw :)

I don't think comparing `td-shim` with `qboot` makes a lot of sense, as @jyao1 answers hopefully clarified. The `shim` part in `td-shim` is relative to a full featured, ovmf based firmware for TDX.
I think the question here would be: Can we remove something from this piece of code and still do `direct` kernel boot on top of TDX? And I believe there's not much to remove.